### PR TITLE
Feature docker client

### PR DESCRIPTION
**Pull request by tushar-tushar**

### CODE REVIEW

File: commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java [hunk 1]
```diff
@@ -200,6 +201,12 @@ public String dockerImageByNameFilterByImageOwner(String imageName) {
         StringBuilder sb = new StringBuilder(dockerImages());
         return sb.append(SPACE).append(imageName).append(filter(IMAGE_OWNER, HYSCALE)).append(quiet()).toString();
     }
+    
+    public Map<String, String> getImageLabelMap(){
+        Map<String, String> labels = new HashMap<>();
+        labels.put(IMAGE_OWNER, HYSCALE);
+        return labels;
+    }
```
Comment (SameeraGrandhi) [inline commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java:209]: Get this out in a different class, feels like it is not part of command but to an Image.
The contract has changed

File: commons/src/main/java/io/hyscale/commons/logger/WorkflowLogger.java [hunk 2]
```diff
@@ -221,5 +221,9 @@ public static void logPersistedActivities() {
         }
         persistedActivities.clear();
     }
-    
+
+    public static void write(String message) {
```
Comment (SameeraGrandhi) [inline commons/src/main/java/io/hyscale/commons/logger/WorkflowLogger.java:225]: Method name is out of sync I feel!
What do you think about this @tushar-tushar ?

File: image-builder/builder-core/src/main/java/io/hyscale/builder/core/models/BuildContext.java [hunk 3]
```diff
@@ -84,8 +85,16 @@ public void setImageRegistry(ImageRegistry imageRegistry) {
 		this.imageRegistry = imageRegistry;
 	}
 
+	public ImageRegistry getPullImageRegistry() {
```
Comment (SameeraGrandhi) [inline image-builder/builder-core/src/main/java/io/hyscale/builder/core/models/BuildContext.java:88]: PullRegistry is commonly used . I think that makes more sense.
The other image registry you can change it PushRegistry

File: image-builder/builder-services/src/main/java/io/hyscale/builder/cleanup/services/impl/DeleteAfterBuildBinary.java [hunk 4]
```diff
@@ -15,56 +15,66 @@
...
- * <p>
+ * Class provides a docker binary based implementation of {@link DeleteAfterBuild}
+ * 
  * docker rmi $(docker images <serviceimage> --filter label=imageowner=hyscale -q)
  */
 
 @Component
-public class DeleteAfterBuild implements ImageCleanupProcessor {
+@Conditional(DockerBinaryCondition.class)
+public class DeleteAfterBuildBinary extends DeleteAfterBuild {
```
Comment (SameeraGrandhi) [inline image-builder/builder-services/src/main/java/io/hyscale/builder/cleanup/services/impl/DeleteAfterBuildBinary.java:45]: Instead of this there should be DockerClient interface to push , pull, tag, build,remove etc. Use these operations in ImageBuilder and in the Cleanup Processor also.

Comment (SameeraGrandhi) [reply image-builder/builder-services/src/main/java/io/hyscale/builder/cleanup/services/impl/DeleteAfterBuildBinary.java:45]: Impl of DockerClient will be DockerBinary and DockerRestClient . They should operate on high level operations of just push, pull, build, tag etc.

File: commons/src/main/java/io/hyscale/commons/logger/WorkflowLogger.java (hunk 2 above)
Comment (tushar-tushar) [reply commons/src/main/java/io/hyscale/commons/logger/WorkflowLogger.java:225]: I can rename it as log keeping it in sync with other method names?

File: image-builder/builder-services/src/main/java/io/hyscale/builder/cleanup/services/impl/DeleteAfterBuildBinary.java (hunk 4 above)
Comment (tushar-tushar) [reply image-builder/builder-services/src/main/java/io/hyscale/builder/cleanup/services/impl/DeleteAfterBuildBinary.java:45]: We have a LocalImageBuildPushServiceImpl which will replace the ImageBuilder, there I can use the DockerClient to perform build, push, pull, etc?
For Image clean up, the docker client will have the remove option, but to get the imageIds to remove, we need different implementation of cleanUp policy which again would be based on Binary and REST client, is this approach fine?

File: commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java [hunk 5]
```diff
@@ -159,7 +146,7 @@ private String docker() {
         return DOCKER_COMMAND + SPACE;
     }
 
-    private String normalize(String input) {
+    public String normalize(String input) {
```
Comment (SameeraGrandhi) [inline commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java:149]: When you are making this public and this method is not related to imagecommandprovider get this out to something like "ImageNameNormalizer"

File: commons/src/main/java/io/hyscale/commons/utils/ImageDetailsUtil.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import io.hyscale.commons.commands.provider.ImageCommandProvider;
+import io.hyscale.commons.constants.ToolConstants;
+
+@Component
+public class ImageDetailsUtil {
```
Comment (SameeraGrandhi) [inline commons/src/main/java/io/hyscale/commons/utils/ImageDetailsUtil.java:29]: 1. This util is giving broad scope but the scope of the class is image metadata only.
2.ImageCommandProvider does not have to depend on this ...if we can see the normalize separately.

File: image-builder/builder-core/src/main/java/io/hyscale/builder/core/models/BuildContext.java [hunk 7]
```diff
@@ -24,8 +24,8 @@
 
 	private DockerfileEntity dockerfileEntity;
 	private DockerImage dockerImage;
-	private ImageRegistry imageRegistry;
-	private ImageRegistry pullImageRegistry;
+	private ImageRegistry pushRegistry;
+	private ImageRegistry pullRegistry;
```
Comment (SameeraGrandhi) [inline image-builder/builder-core/src/main/java/io/hyscale/builder/core/models/BuildContext.java:28]: Give some description to this registries.

File: commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java [hunk 8]
```diff
@@ -44,29 +49,15 @@
...
-        StringBuilder sb = new StringBuilder();
-        sb.append(HYSCALE_IO_URL).append(SLASH).append(appName).append(SLASH).append(serviceName);
-        if (StringUtils.isNotBlank(tag)) {
-            sb.append(ToolConstants.COLON).append(tag);
-        }
-        return normalize(sb.toString());
-    }
+	
+	@Autowired
+	private ImageMetadataUtil imageMetadataUtil;
```
Comment (SameeraGrandhi) [inline commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java:60]: This name request has not been resolved. 
Provide a better name as referred to in the previous comments.

Comment (tushar-tushar) [reply commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java:60]: As per the previous comment, the name should reflect the scope of the class which is restricted to the Image metadata, hence the name ImageMetadataUtil. Any better suggestions?

Comment (SameeraGrandhi) [reply commons/src/main/java/io/hyscale/commons/commands/provider/ImageCommandProvider.java:60]: I would suggest "ImageMetadataProvider".
I don't see this class as a utility. What do you say?
